ssl/quic/quic_port.c

```diff
@@ -739,6 +759,130 @@ static int marshal_validation_token(QUIC_VALIDATION_TOKEN *token,
+                                    size_t pt_len,
+                                    unsigned char *ciphertext,
+                                    size_t *ct_len)
+{
+    EVP_CIPHER_CTX *ctx = NULL;
+    int len, ret = 0;
+    EVP_CIPHER *cipher = NULL;
+    size_t iv_len, tag_len;
+    unsigned char *iv = ciphertext, *data, *tag;
+
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL
+        || (cipher = EVP_CIPHER_fetch(port->engine->libctx,
+                                      "AES-256-GCM", NULL)) == NULL
+        || !EVP_EncryptInit_ex(ctx, cipher, NULL, NULL, NULL)
+        || (tag_len = EVP_CIPHER_CTX_get_tag_length(ctx)) == 0
+        || (iv_len = EVP_CIPHER_get_iv_length(cipher)) == 0)
+        goto err;
+
+    *ct_len = iv_len + pt_len + tag_len + QUIC_RETRY_INTEGRITY_TAG_LEN;
+    if (ciphertext == NULL) {
+        ret = 1;
+        goto err;
+    }
+
+    data = ciphertext + iv_len;
+    tag = data + pt_len;
+
+    if (!RAND_bytes_ex(port->engine->libctx, ciphertext, iv_len, 0)
+        || !EVP_EncryptInit_ex(ctx, NULL, NULL, port->token_key, iv)
+        || !EVP_EncryptUpdate(ctx, data, &len, plaintext, pt_len)
+        || !EVP_EncryptFinal_ex(ctx, data + pt_len, &len)
+        || !EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, tag_len, tag))
+        goto err;
+
+    ret = 1;
+err:
+    EVP_CIPHER_CTX_free(ctx);
+    EVP_CIPHER_free(cipher);
+    return ret;
+}
+
+/**
+ * @brief Decrypts a validation token using AES-256-GCM
+ *
+ * @param port       The QUIC port containing the decryption key
+ * @param ciphertext The encrypted data (including IV and tag)
+ * @param ct_len     Length of the ciphertext
+ * @param plaintext  Buffer to receive decrypted data. If NULL, pt_len will be
+ *                   set to the required buffer size.
+ * @param pt_len     Pointer to size_t that will receive the plaintext length
+ *
+ * @return 1 on success, 0 on failure
+ *
+ * Expected ciphertext format:
+ * [EVP_GCM_IV_LEN bytes IV][encrypted data][EVP_GCM_TAG_LEN bytes tag]
+ */
+static int decrypt_validation_token(const QUIC_PORT *port,
+                                    const unsigned char *ciphertext,
+                                    size_t ct_len,
+                                    unsigned char *plaintext,
+                                    size_t *pt_len)
+{
+    EVP_CIPHER_CTX *ctx = NULL;
+    EVP_CIPHER *cipher = NULL;
+    int len = 0, ret = 0;
+    size_t iv_len, tag_len;
+    const unsigned char *iv = ciphertext, *data, *tag;
+
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL
+        || (cipher = EVP_CIPHER_fetch(port->engine->libctx,
+                                      "AES-256-GCM", NULL)) == NULL
+        || !EVP_DecryptInit_ex(ctx, cipher, NULL, NULL, NULL)
+        || (tag_len = EVP_CIPHER_CTX_get_tag_length(ctx)) == 0
+        || (iv_len = EVP_CIPHER_get_iv_length(cipher)) == 0)
+        goto err;
+
+    /* Prevent malloc of a buffer that is too large */
+    if (ct_len < iv_len + tag_len || ct_len > 512)
```

Comment:
Can you elaborate on why 512 is the upper bound of ct_len here?  Is that just an arbitrary upper bound to prevent a malicious attacker from forcing a large allocation?

Reply:
Just an arbitrary value that I thought should be more than sufficient for current and futures changes to the validation token while preventing large allocations (since it’s a client-provided value). All the max sizes in the validation token are calculable except remote address size, but I assume that can only be like max 20-ish bytes.

If you have a better idea of how to calculate the upper bound of the size of the token, please let me know your thoughts.

Reply:
Looking at the current validation token in the worst case:
* timestamp is 8 bytes
* odcid and rscid are maximally 42 bytes in total
* remote_addr_len is a size_t (8 bytes)
* remote_addr is in the worst case 110 bytes (in the case of using a maximally sized AF_UNIX socket)
* is_retry is a single byte
So in the worst case, we're looking at an unencrypted token of 169 bytes.  AES-GCM IIRC has a 128 byte block size so after encryption we're looking at 256 bytes, plus a 16 byte tag, and I think a 12 byte IV, so in total an opaque 284 byte blob to send.  

Thats not close, but maybe in striking distance if we  add more than 215 more bytes to the pre-encrypted value .  Thats probably enough space to stay safe for now.

I wouldn't advocate for it immediately, but if you're making other changes and were so inclined, you might could do something like this:
1) Create a per-port variable max_ct_len, initalized to zero
2) On token_encrypt, compare the computed ct_len to port->max_ct_len.  If ct_len is greater than port->max_ct_len, store the new, larger value of ct_len to port->max_ct_len.
3) On token decode check to see if the passed in ct_len is greater than port->max_ct_len.

A failed test in step 3 would indicate you got a token larger than the largest one this port has ever encoded, which would be impossible, and cause to raise an error.

Its a bit of work, and might be nice as it future proofs token expansion, but if you've got other stuff to do, maybe just make a comment above the 512 indicating this is an estimated value and something to adjust as the token expands in the future

Reply:
> * AES-GCM IIRC has a 128 byte block size so after encryption we're looking at 256 bytes

The block size for AES is 16 bytes (128 bits) however AES-GCM is a CTR mode cipher so there is no block size limitation. So 28 bytes for tag & iv + 169 bytes plaintext size gives just 197 bytes. We could and should use a fixed stack-allocated buffer. There is no point in doing malloc for this.

Reply:
Ok, I will also statically allocate the buffer for marshalling the token